File: csharp/palindrome-products/PalindromeProducts.cs
using System.Net.NetworkInformation;

public static class PalindromeProducts
{
    public static (int, IEnumerable<(int, int)>) Largest(int minFactor, int maxFactor)
    {
        IEnumerable<int> results = GetPalindromes(minFactor, maxFactor);
        if (results.Count() == 0)
            throw new ArgumentException($"No factors found for range of ({minFactor}, {maxFactor})");
        int largest = results.Max();
        return (largest, Factors(largest, minFactor, maxFactor));
    }

    public static (int, IEnumerable<(int, int)>) Smallest(int minFactor, int maxFactor)
    {
        IEnumerable<int> results = GetPalindromes(minFactor, maxFactor);
        if (results.Count() == 0)
            throw new ArgumentException($"No factors found for range of ({minFactor}, {maxFactor})");
        int smallest = results.Min();
        return (smallest, Factors(smallest, minFactor, maxFactor));
    }

    public static IEnumerable<int> GetPalindromes(int minFactor, int maxFactor)
    {
        if (minFactor > maxFactor)
            throw new ArgumentException($"Range of ({minFactor}, {maxFactor}) is not valid.");

        for (int j = minFactor; j <= maxFactor; j++)
        {
            for (int i = minFactor; i <= maxFactor; i++)
            {
                if (i > j) // no use computing this twice.
                    continue;

                if (i * j == 0) // Zero doesn't count.
                    continue;

                if (IsPalindrome(i * j))
                    yield return (i * j);
            }

        }
    }

    public static bool IsPalindrome(int value)
    {
        string num = value.ToString();
        int start = 0;
        int end = num.Length - 1;
        while (start < end)
        {
            if (num[start] != num[end])
                return false;
            start++;
            end--;
        }
        return true;
    }

    public static IEnumerable<(int, int)> Factors(int value, int minFactor, int maxFactor)
    {
        for (int j = minFactor; j <= maxFactor; j++)
        {
            for (int i = minFactor; i <= maxFactor; i++)
            {
                if (j > i)
                    continue; // Only want each factor pair once.
                if ((i * j) == 0) // Zero doesn't count.
                    continue;
                if (i > value || j > value) // Won't be a factor if larger than value.
                    continue;
                if ((i * j) == value)
                    yield return (j, i);
            }
        }
    }
}

File: go/palindrome-products/palindrome_products.go
// Palindrom Products exercise
package palindromeproducts

import (
	"errors"
	"strconv"
)

// Define Product type here.
type Product struct {
	Product        int
	Factorizations [][2]int
}

// Create a new Product struct with default values
// @returns: Product struct initialized with defaults (Product 0, empty Factorizations)
func NewProduct() Product {
	return Product{Product: 0, Factorizations: [][2]int{}}
}

// Find the palindrome products between fmin and fmax, return the smallest and largest
// @param fmin: The start of the number range
// @param fmax: The end of the number range (inclusize)
// @returns: First Product is minimum Palindrome product, second is the Maximum. Will return Empty/New Products if none found.
// @raises: Error if fmin > fmax
func Products(fmin, fmax int) (Product, Product, error) {

	if fmin > fmax {
		return NewProduct(), NewProduct(), errors.New("min must be <= max")
	}
	candidates := make(map[int][][2]int, 0)
	for j := fmin; j <= fmax; j++ {
		for i := j; i <= fmax; i++ {
			product := i * j
			if IsPalindrome(product) {
				factors, ok := candidates[product]
				if !ok {
					factors = make([][2]int, 0)
				}
				found := false
				for _, factor := range factors {
					if factor[0] == j && factor[i] == j {
						found = true
						break
					}
				}
				if !found {
					factors = append(factors, [2]int{j, i})
				}
				candidates[product] = factors
			}
		}
	}
	if len(candidates) == 0 {
		return NewProduct(), NewProduct(), nil
	}

	first := true
	var min Product = NewProduct()
	var max Product = NewProduct()
	for key, value := range candidates {
		if first || key < min.Product {
			min.Product = key
			min.Factorizations = value
		}

		if first || key > max.Product {
			max.Product = key
			max.Factorizations = value
		}

		first = false
	}
	return min, max, nil
}

// Check if a number is a Palindrome
// @param num: The number to check
// @returns: True if a palindrome (written the same forward and reversed)
func IsPalindrome(num int) bool {
	value := []rune(strconv.Itoa(num))
	begin := 0
	end := len(value) - 1
	for begin < end {
		if value[begin] != value[end] {
			return false
		}
		begin++
		end--
	}
	return true
}
